lib/compat/wordpress-6.5/blocks.php

```diff
@@ -65,68 +65,22 @@ function gutenberg_block_bindings_replace_html( $block_content, $block_name, str
-			// The button block has a wrapper while the paragraph and heading blocks don't.
-			if ( 'core/button' === $block_name ) {
-				$button_wrapper                 = $block_reader->get_tag();
-				$button_wrapper_attribute_names = $block_reader->get_attribute_names_with_prefix( '' );
-				$button_wrapper_attrs           = array();
-				foreach ( $button_wrapper_attribute_names as $name ) {
-					$button_wrapper_attrs[ $name ] = $block_reader->get_attribute( $name );
-				}
+			// Hardcode the selectors and processing until the HTML API is able to read CSS selectors and replace inner HTML.
+			// TODO: Use the HTML API instead.
+			if ( 'core/paragraph' === $block_name && 'content' === $attribute_name ) {
+				$selector = 'p';
 			}
-
-			foreach ( $selectors as $selector ) {
-				// If the parent tag, or any of its children, matches the selector, replace the HTML.
-				if ( strcasecmp( $block_reader->get_tag( $selector ), $selector ) === 0 || $block_reader->next_tag(
-					array(
-						'tag_name' => $selector,
-					)
-				) ) {
-					$block_reader->release_bookmark( 'iterate-selectors' );
-
-					// TODO: Use `set_inner_html` method whenever it's ready in the HTML API.
-					// Until then, it is hardcoded for the paragraph, heading, and button blocks.
-					// Store the tag and its attributes to be able to restore them later.
-					$selector_attribute_names = $block_reader->get_attribute_names_with_prefix( '' );
-					$selector_attrs           = array();
-					foreach ( $selector_attribute_names as $name ) {
-						$selector_attrs[ $name ] = $block_reader->get_attribute( $name );
-					}
-					$selector_markup = "<$selector>" . wp_kses_post( $source_value ) . "</$selector>";
-					$amended_content = new WP_HTML_Tag_Processor( $selector_markup );
-					$amended_content->next_tag();
-					foreach ( $selector_attrs as $attribute_key => $attribute_value ) {
-						$amended_content->set_attribute( $attribute_key, $attribute_value );
-					}
-					if ( 'core/paragraph' === $block_name || 'core/heading' === $block_name ) {
-						return $amended_content->get_updated_html();
-					}
-					if ( 'core/button' === $block_name ) {
-						$button_markup  = "<$button_wrapper>{$amended_content->get_updated_html()}</$button_wrapper>";
-						$amended_button = new WP_HTML_Tag_Processor( $button_markup );
-						$amended_button->next_tag();
-						foreach ( $button_wrapper_attrs as $attribute_key => $attribute_value ) {
-							$amended_button->set_attribute( $attribute_key, $attribute_value );
-						}
-						return $amended_button->get_updated_html();
-					}
-				} else {
-					$block_reader->seek( 'iterate-selectors' );
-				}
+			if ( 'core/heading' === $block_name && 'content' === $attribute_name ) {
+				$selector = 'h[1-6]';
 			}
-			$block_reader->release_bookmark( 'iterate-selectors' );
-			return $block_content;
+			if ( 'core/button' === $block_name && 'text' === $attribute_name ) {
+				$selector = 'a';
```

Comment:
I think a button block can also be a `<button>` element depending on the value of the `tagName` attribute.

Reply:
You're right it seems supported according to the attributes. Although I checked how that attribute works and I couldn't find a way to use a `<button>` instead of a link tag.

I added [this commit](https://github.com/WordPress/gutenberg/pull/61236/commits/57c2ce1723f66cd6db8f86c6007d21bef9ddc211) to cover that possibility.